File: Tree/Medium/103. Binary Tree Zigzag Level Order Traversal - Iterative.java
package Tree.Medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
    
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {return result;}
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int depth = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> subList = new LinkedList<>();
            for (int i=0; i<size; ++i) {
                TreeNode node = queue.remove();
                if (depth % 2 == 0) {
                    subList.add(node.val);
                } else {
                    subList.add(0, node.val);
                }
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
            depth++;
            result.add(subList);
        }
        return result;
    }

}
        
        

File: Tree/Medium/103. Binary Tree Zigzag Level Order Traversal - Iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue = [root]
        depth = 0
        result = []
        while queue:
            length = len(queue)
            sub_queue = []
            for i in range(length):
                node = queue.pop(0)
                if (depth % 2 == 0):
                    sub_queue.append(node.val)
                else:
                    sub_queue.insert(0, node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            depth += 1
            result.append(sub_queue)
        return result
                
        